src/EventuallyQueue.js

```diff
@@ -0,0 +1,188 @@
+        }
+      }
+    } else {
+      index = queueData.length;
+    }
+    queueData[index] = {
+      queueId,
+      action,
+      object,
+      serverOptions,
+      id: object.id,
+      className: object.className,
+      hash: object.get('hash'),
+      createdAt: new Date(),
+    };
+    return this.setQueue(queueData);
+  },
+
+  async getQueue() {
+    const q = await Storage.getItemAsync(this.localStorageKey);
+    if (!q) {
+      return [];
+    }
+    return JSON.parse(q);
+  },
+
+  setQueue(queueData) {
+    return Storage.setItemAsync(this.localStorageKey, JSON.stringify(queueData));
+  },
+
+  async remove(queueId) {
+    const queueData = await this.getQueue();
+    const index = this.queueItemExists(queueData, queueId);
+    if (index > -1) {
+      queueData.splice(index, 1);
+      await this.setQueue(queueData);
+    }
+  },
+
+  clear() {
+    return Storage.setItemAsync(this.localStorageKey, JSON.stringify([]));
+  },
+
+  queueItemExists(queueData, queueId) {
+    return queueData.findIndex(data => data.queueId === queueId);
+  },
+
+  async length() {
+    const queueData = await this.getQueue();
+    return queueData.length;
+  },
+
+  async sendQueue() {
+    const queueData = await this.getQueue();
+    if (queueData.length === 0) {
+      return false;
+    }
+    for (let i = 0; i < queueData.length; i += 1) {
+      const ObjectType = ParseObject.extend(queueData[i].className);
+      if (queueData[i].id) {
+        await this.reprocess.byId(ObjectType, queueData[i]);
+      } else if (queueData[i].hash) {
+        await this.reprocess.byHash(ObjectType, queueData[i]);
+      } else {
+        await this.reprocess.create(ObjectType, queueData[i]);
+      }
+    }
+    return true;
+  },
+
+  async sendQueueCallback(object, queueObject) {
+    if (!object) {
+      return this.remove(queueObject.queueId);
+    }
+    switch (queueObject.action) {
+    case 'save':
+      // Queued update was overwritten by other request. Do not save
+      if (
+        typeof object.updatedAt !== 'undefined' &&
+          object.updatedAt > new Date(queueObject.object.createdAt)
+      ) {
+        return this.remove(queueObject.queueId);
+      }
+      try {
+        await object.save(queueObject.object, queueObject.serverOptions);
+        await this.remove(queueObject.queueId);
+      } catch (e) {
+        if (e.message !== 'XMLHttpRequest failed: "Unable to connect to the Parse API"') {
+          await this.remove(queueObject.queueId);
+        }
+      }
+      break;
+    case 'destroy':
+      try {
+        await object.destroy(queueObject.serverOptions);
+        await this.remove(queueObject.queueId);
+      } catch (e) {
+        if (e.message !== 'XMLHttpRequest failed: "Unable to connect to the Parse API"') {
+          await this.remove(queueObject.queueId);
+        }
+      }
+      break;
+    }
+  },
+
+  poll() {
+    if (this.polling) {
```

Comment:
Could this cause race conditions issues?

Reply:
Good question. I'll have to think on it.

Reply:
By raise condition do you mean that multiple polling could happen at the same time?

Reply:
Yes, as in calling `poll` multiple times from different threads, then `this.polling` may not be set thread safely.

Reply:
I was thinking about changing the setInterval to setTimeout implementation.

Reply:
I think that makes sense since we're looking at a network op. For example, if the network op times out only after 60 seconds (or more), and the setInterval also executes every 60 seconds (or less), there may be an overlap. That issue wouldn't exists using a nested setTimeout.

Reply:
I think that makes sense since we're looking at a network op. For example, if the network op times out only after 60 seconds, and the setInterval also executes every 61 seconds, then the time between network retries would be only 1 second. And maybe in the previous 60 seconds, the host system also has some retry mechanism before it times out, so it would be constantly trying to reach the server. That issue wouldn't exists using a nested setTimeout.

Technically, polling could hardly ever overlap in JS (except for edge cases with multi-worker maybe), so the if clause there should be fine whether setTimeout or setInterval is used, both are based on the same timer mechanism.

Reply:
I fixed the raise condition issue, just to be sure.